src/vs/workbench/contrib/chat/browser/chatSessions.contribution.ts

```diff
@@ -877,60 +875,43 @@ export class ChatSessionsService extends Disposable implements IChatSessionsServ
 		};
 	}
 
-	public registerModelProgressListener(model: IChatModel, callback: () => void): void {
-		// Prevent duplicate registrations for the same model
-		if (this._registeredModels.has(model)) {
-			return;
-		}
-		this._registeredModels.add(model);
-
-		// Helper function to register listeners for a request
-		const registerRequestListeners = (request: IChatRequestModel) => {
-			if (!request.response || this._registeredRequestIds.has(request.id)) {
-				return;
-			}
-
-			this._registeredRequestIds.add(request.id);
-
-			this._register(request.response.onDidChange(() => {
-				callback();
-			}));
+	public registerChatModelChangeListeners(
+		chatService: IChatService,
+		chatSessionType: string,
+		onChange: () => void
+	): IDisposable {
+		const chatModelsICareAbout = chatService.chatModels.map(models =>
+			Array.from(models).filter((model: IChatModel) => model.sessionResource.scheme === chatSessionType)
+		);
 
-			// Track tool invocation state changes
-			const responseParts = request.response.response.value;
-			responseParts.forEach((part: IChatProgressResponseContent) => {
-				if (part.kind === 'toolInvocation') {
-					const toolInvocation = part as IChatToolInvocation;
-					// Use autorun to listen for state changes
-					this._register(autorunSelfDisposable(reader => {
-						const state = toolInvocation.state.read(reader);
+		const listeners = new ResourceMap<IDisposable>();
```

Comment:
I think we actually want to return a DisposableStore containing both the `autorunIterableDelta` and something like

```ts
store.add(toDisposable(() => {
  for (const listener of listeners) { listener.dispose(); }
}));
```

otherwise they might get orphaned